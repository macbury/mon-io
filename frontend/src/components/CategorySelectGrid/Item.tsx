import React, { useCallback } from 'react'
import { Text, TouchableRipple } from 'react-native-paper'
import styled, { DefaultTheme } from 'styled-components/native'
import CategoryIcon from '../CategoryIcon'
import { IItemProps } from './types'

interface IItemContainerProps {
  height: number;
  selected: boolean;
  theme: DefaultTheme;
}

const ItemContainer = styled(TouchableRipple)`
  height: ${(props : IItemContainerProps) => props.height};
  width: ${(props : IItemContainerProps) => props.height};
  flex-flow: row wrap;
  border-width: 1px;
  opacity: ${(props : IItemContainerProps) => props.selected ? 1.0 : 0.8};
  border-color: ${(props : IItemContainerProps) => props.selected ? props.theme.itemSelectedBorderColor : props.theme.itemBorderColor};
`

const ItemInner = styled.View`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  flex: 1;
`

interface INameTextProps {
  theme?: DefaultTheme;
  color?: string;
}

const NameText = styled(Text)`
  color: ${(prop : INameTextProps) => prop.color ? prop.color : prop.theme.colors.text};
`

export default function Item({ category, size, selected, onSelectCategory } : IItemProps) {
  const onItemPress = useCallback(() => {
    onSelectCategory(category)
  }, [category, onSelectCategory])

  return (
    <ItemContainer selected={selected} height={size} onPress={onItemPress} rippleColor={`${category.color}3C`}>
      <ItemInner>
        {category.icon && <CategoryIcon color={category.color} name={category.icon} />}
        <NameText color={category.color}>
          {category.name}
        </NameText>
      </ItemInner>
    </ItemContainer>
  )
}