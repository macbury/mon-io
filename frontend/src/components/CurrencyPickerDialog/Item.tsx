import React, { useCallback } from 'react'
import styled from 'styled-components/native'
import { RadioButton, Text, TouchableRipple } from 'react-native-paper'
import { Currency } from '../../api/graphql'
import { IOnCurrencySelect } from './shared'

const Inner = styled.View`
  padding: 10px 20px;
  flex-direction: row;
  align-items: center;
`

const Name = styled(Text)`
  flex: 1;
  padding-right: 15px;
`

const Symbol = styled(Text)`
  opacity: 0.4;
  font-weight: bold;
`

interface IItemProps extends IOnCurrencySelect {
  currency: Currency
  selected?: boolean
}

export default function Item({ currency, selected, onCurrencySelect } : IItemProps) {
  const onCurrencyPress = useCallback(() => onCurrencySelect(currency), [currency, onCurrencySelect])

  return (
    <TouchableRipple onPress={onCurrencyPress}>
      <Inner>
        <RadioButton
          value="first"
          status={selected ? 'checked' : 'unchecked'}
          onPress={onCurrencyPress}
        />
        <Name numberOfLines={1}>{currency.name}</Name>
        <Symbol>{currency.symbol}</Symbol>
      </Inner>
    </TouchableRipple>
  )
}