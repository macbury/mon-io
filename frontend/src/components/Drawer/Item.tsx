import React, { useCallback } from 'react'
import { useNavigation } from 'react-navigation-hooks'
import { TouchableOpacity } from 'react-native'
import { NavigationNavigateAction, StackActions } from 'react-navigation'
import { List, Text } from 'react-native-paper'
import styled, { DefaultTheme, useTheme } from 'styled-components/native'
import Link from '../Link'

import { useTranslation } from 'react-i18next'

interface IStyleProps {
  collapsed?: boolean
  selected?: boolean
  theme?: DefaultTheme
}

const ItemLink = styled(Link)`

`

const Content = styled.View`
  flex-direction: row;
  align-items: center;
  margin: 5px;
  padding: ${({ collapsed } : IStyleProps) => collapsed ? '0px' : '5px'};
  background: ${({ theme, selected } : IStyleProps) => selected ? theme.colors.primary+'15' : 'transparent'};
  border-radius: 5px;
`

const Icon = styled(List.Icon)`
  flex: 1;
  min-width: 40px;
  max-width: ${({ collapsed } : IStyleProps) => collapsed ? '100%' : '40px'};
  height: 40px;
  padding: 2px;
  margin: 0px;
`

const Label = styled(Text)`
  font-size: 16px;
  flex: 1;
  margin-left: 10px;
  color: ${({ theme, selected } : IStyleProps) => selected ? theme.colors.primary : theme.colors.text};
`

export interface IItemProps {
  collapsed?: boolean;
  selected?: boolean;
  /** I18n key */
  titleKey: string;
  icon: string;
  action?: NavigationNavigateAction;
  onPress?();
}

export default function Item({ selected, collapsed, action, titleKey, onPress, icon } : IItemProps) {
  const { t } = useTranslation()
  const theme = useTheme()
  const navigation = useNavigation()

  const Wrapper : any = !action || selected ? TouchableOpacity : ItemLink

  const executePressLogic = useCallback(() => {
    if (action && selected) {
      navigation.dispatch(StackActions.popToTop())
    } else if (onPress) {
      onPress()
    }
  }, [onPress, action, selected, navigation])

  return (
    <Wrapper ripple={false} action={action} onPress={executePressLogic} title={t(titleKey)}>
      <Content selected={selected}>
        <Icon
          color={selected ? theme.colors.primary : theme.colors.text}
          collapsed={collapsed}
          icon={icon} />
        {!collapsed && <Label selected={selected}>{t(titleKey)}</Label>}
      </Content>
    </Wrapper>
  )
}