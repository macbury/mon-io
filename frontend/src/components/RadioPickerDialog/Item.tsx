import React, { useCallback } from 'react'
import styled from 'styled-components/native'
import { RadioButton, Text, TouchableRipple } from 'react-native-paper'
import { RadioOption } from '../../stores/UI/RadioPickerDialogStore'

const Inner = styled.View`
  padding: 10px 20px;
  flex-direction: row;
  align-items: center;
`

const Name = styled(Text)`
  flex: 1;
  padding-right: 15px;
`

const Symbol = styled(Text)`
  opacity: 0.4;
  font-weight: bold;
`

interface IItemProps{
  option: RadioOption
  selected?: boolean
  onSelect(option: RadioOption)
}

export default function Item({ option, selected, onSelect } : IItemProps) {
  const onOptionPress = useCallback(() => onSelect(option), [option, onSelect])

  return (
    <TouchableRipple onPress={onOptionPress}>
      <Inner>
        <RadioButton
          value="first"
          status={selected ? 'checked' : 'unchecked'}
          onPress={onOptionPress}
        />
        <Name numberOfLines={1}>{option.title}</Name>
      </Inner>
    </TouchableRipple>
  )
}