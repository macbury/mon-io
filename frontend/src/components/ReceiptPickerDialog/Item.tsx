import React, { useMemo } from 'react'
import { TouchableRipple, Text } from 'react-native-paper'
import moment from 'moment-timezone'
import { Receipt } from '../../api/graphql'
import FileTypeIcon from '../Receipts/FileTypeIcon'
import styled from 'styled-components/native'

const ItemContainer = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
`

const ItemLink = styled(TouchableRipple)`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 10px 25px 10px 20px;
  flex-shrink: 1;
  flex-grow: 1;
  border-bottom-width: 1px;
  border-bottom-color: ${(props) => props.theme.itemBorderColor};
`

const ItemCover = styled.View`
  width: 50px;
  height: 50px;
  margin-right: 15px;
  overflow: hidden;
  display: flex;
  align-items: center;
`

const ItemDetails = styled.View`
  flex: 1;
  flex-shrink: 1;
`

const PublishedAtText = styled(Text)`
  opacity: 0.4;
  text-transform: uppercase;
`

const TitleText = styled(Text)`
  font-weight: bold;
`

interface ICategoryTextProp {
  color: string;
}

const CategoryText = styled(Text)`
  color: ${(prop : ICategoryTextProp) => prop.color ? prop.color : '#fff'};
`

export interface IItemProps {
  receipt: Receipt
  onReceiptSelect(receipt : Receipt)
}

export default function Item(props : IItemProps) {
  const {
    receipt,
    onReceiptSelect
  } = props
  const publishedAt = useMemo(() => moment(receipt.createdAt).format('D/MM/YYYY H:mm'), [receipt.createdAt])

  return (
    <ItemLink onPress={() => onReceiptSelect(receipt)}>
      <ItemContainer>
        <ItemCover>
          <FileTypeIcon type={receipt.mimeType} size={48} />
        </ItemCover>
        <ItemDetails>
          <PublishedAtText>{publishedAt}</PublishedAtText>
          <TitleText lineBreakMode="tail" numberOfLines={2}>{receipt.name}</TitleText>
          {receipt.category && <CategoryText color={receipt.category.color}>in {receipt.category.name}</CategoryText>}
        </ItemDetails>
      </ItemContainer>
    </ItemLink>
  )
}