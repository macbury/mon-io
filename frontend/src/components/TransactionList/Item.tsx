import React, { useMemo } from 'react'
import moment from 'moment-timezone'
import styled, { useTheme, DefaultTheme } from 'styled-components/native'

import { useTranslation } from 'react-i18next'
import { View } from 'react-native'
import { Text } from 'react-native-paper'
import { NavigationNavigateAction } from 'react-navigation'

import { Transaction, Recurrence } from '../../api/graphql'
import CategoryIcon from '../CategoryIcon'
import Link from '../Link'
import Currency from '../Currency'
import { useItemActionSheet } from '../TransactionForm/EditMenu'
import { DetailsIcon, FaviconIcon, DETAILS_ICON_SIZE } from './DetailsIcons'

interface ITransactionItemProps {
  transaction: Transaction
  simple? : boolean
  style? : any
  action: NavigationNavigateAction
}

const Container = styled.View`
  display: flex;
  flex-direction: row;
  padding: 10px 15px;
  align-content: center;
  border-bottom-width: 1px;
  border-bottom-color: ${(prop) => prop.theme.itemBorderColor};
  align-items: center;
`

const Details = styled.View`
  display: flex;
  flex-direction: column;
  flex: 1;
`

const Icon = styled(CategoryIcon)`
  margin-bottom: 0px;
  margin-right: 15px;
`

interface ICategoryNameProps {
  theme: DefaultTheme
  color: string
}

const CategoryName = styled(Text)`
  font-weight: bold;
  font-size: 16px;
  color: ${({ color } : ICategoryNameProps) => color};
`

const Date = styled(Text)`
  opacity: 0.4;
  margin-right: 10px;
`

const DetailsRow = styled.View`
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  margin-top: 3px;
`

const DetailsLocation = styled(Text)`
  margin-left: 5px;
`

const Notes = styled(Text)`
  font-style: italic;
  flex: 1;
`

function LinkToTransaction({ transaction, action, children, ...props }) {
  const showActionSheet = useItemActionSheet({ transaction })

  return (
    <Link ripple onLongPress={showActionSheet} action={action} {...props}>
      {children}
    </Link>
  )
}

function ContainerForTransaction({ children, ...props }) {
  return (
    <View {...props}>
      {children}
    </View>
  )
}

export default function TransactionItem({ transaction, simple, ...props } : ITransactionItemProps) {
  const theme = useTheme()
  const date = useMemo(() => (moment(transaction.date).format('Do MMM YYYY HH:MM')), [transaction.date])
  const { t } = useTranslation()

  const Wrapper = simple ? ContainerForTransaction : LinkToTransaction

  return (
    <Wrapper transaction={transaction} {...props}>
      <Container>
        <Icon
          simple={simple}
          name={transaction.category.icon}
          color={transaction.category.color}
          size={48} />
        <Details>
          <DetailsRow>
            <CategoryName color={transaction.category.color}>{transaction.category.name}</CategoryName>
            {transaction.location && <DetailsLocation numberOfLines={1} ellipsizeMode='head'>{t('pages.summary.transactions.item.at_location', { location: transaction.location.name })} <DetailsIcon name="map" color={theme.colors.text} size={DETAILS_ICON_SIZE} /></DetailsLocation>}
          </DetailsRow>

          <DetailsRow>
            <Date>{date}</Date>
            {transaction.receipt && <DetailsIcon name="attachment" color={theme.colors.text} size={DETAILS_ICON_SIZE} />}
            {transaction.recurrence !== Recurrence.None && <DetailsIcon name="repeat" color={theme.colors.text} size={DETAILS_ICON_SIZE} />}
            {transaction.category.shared && <DetailsIcon name="share-variant" color={theme.colors.text} size={DETAILS_ICON_SIZE} />}
            {transaction.import && <DetailsIcon name="database-import" color={theme.colors.text} size={DETAILS_ICON_SIZE} />}
            <FaviconIcon url={transaction?.link?.faviconUrl} />
            {transaction.notes.length > 0 && <Notes numberOfLines={1}><DetailsIcon name="note" color={theme.colors.text} size={DETAILS_ICON_SIZE} /> {transaction.notes}</Notes>}
          </DetailsRow>
        </Details>
        <Currency
          amount={transaction.amount}
          size={18} />
      </Container>
    </Wrapper>
  )
}