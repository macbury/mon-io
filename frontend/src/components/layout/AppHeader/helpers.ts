import { AppHeaderProp } from './types'

export function getTitleFromNavigation(props : AppHeaderProp) : string {
  let title = props.title
  if (!title) {
    title = props.scene?.descriptor?.options?.title
  }
  return title
}