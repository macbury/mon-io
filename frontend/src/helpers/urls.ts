import { Recurrence } from './../api/graphql';
import moment, { Moment } from 'moment-timezone'
import { NavigationActions, NavigationNavigateAction } from 'react-navigation'
import { TransactionCategoryKind, RecurrenceUpdateMode } from '../api/graphql'

export function homePath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'home' })
}

export function receiptsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'receipts' })
}

export function mapPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'map' })
}

export function authPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function importsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'imports' })
}

export function editImportPath(importId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newImport', params: { importId } })
}

export function settingsApiExplorerPath(refreshTokenId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'apiExplorer', params: { refreshTokenId } })
}

export function categoriesPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'categories' })
}

export function newCategoryPath(kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newCategory', params: { kind } })
}

export function editCategoriesPath(categoryId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategory', params: { categoryId } })
}

export function scanPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'scan' })
}

export function changeNewTransactionCategoryPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'changeNewTransactionCategory' })
}

export function signInPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function seriesPath(date?: Moment) : NavigationNavigateAction {
  const params = date ? { date: date.format('YYYY-MM-DD') } : {}
  return NavigationActions.navigate({ routeName: 'series', params })
}

export function chartsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'charts' })
}

export function editSeriesPath(seriesId : string, updateType: RecurrenceUpdateMode, occuredAt?: Moment | string) : NavigationNavigateAction {
  let date = occuredAt ? moment(occuredAt).format('YYYY-MM-DD') : null
  return NavigationActions.navigate({ routeName: 'editSeries', params: { seriesId, date, updateType } })
}

export function editCategoryBudgetPath(budgetId : string, categoryId : string, kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategoryBudget', params: { budgetId, categoryId, kind } })
}

export function budgetPath(date : Moment) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'budget', params: { date: date.format('YYYY-MM') } })
}

export function summaryPath(date : Moment, type?: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'summary', params: { date: date.format('YYYY-MM'), type } })
}

export function transactionsPath(date? : Moment, type?: TransactionCategoryKind, importId?: string) : NavigationNavigateAction {
  return NavigationActions.navigate({
    routeName: 'transactions',
    params: {
      date: date?.format('YYYY-MM'),
      type,
      importId
    }
  })
}

export function previewReceiptPath(receiptId : string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'preview', params: { receiptId } })
}

export function settingsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settings' })
}

export function settingsRefreshTokensPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsRefreshTokens' })
}

export function settingsAuthorizeDevicePath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsQuickLogin' })
}

export function settingsDownloadMobileAppPath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsDownloadMobileApp' })
}

export function newSummaryTransactionPath(categoryId: string, createdAt: string, kind: TransactionCategoryKind) {
  return NavigationActions.navigate({
    routeName: 'newSummaryTransaction',
    params: {
      categoryId, createdAt, kind
    }
  })
}

export function newBillTransactionPath(receiptId: string) {
  return NavigationActions.navigate({ routeName: 'newBillTransaction', params: { receiptId } })
}

export function newPlannedTransactionPath(seriesId: string, createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newPlannedTransaction', params: { createdAt, seriesId } })
}

export function newSeriesPath(createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newSeries', params: { createdAt, recurrence: Recurrence.Once } })
}

export function editSummaryTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editSummaryTransaction', params: { transactionId } })
}

export function editTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editTransaction', params: { transactionId } })
}