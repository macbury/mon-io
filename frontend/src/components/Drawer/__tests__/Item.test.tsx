import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import { homePath } from '../../../helpers/urls'
import Item, { IItemProps } from '../Item'
import { ThemeProvider } from 'styled-components/native'

jest.mock('react-navigation-hooks', () => ({
  useNavigation: jest.fn()
}))

it('renders', () => {
  const props : IItemProps = {
    icon: 'settings',
    titleKey: 'pages.series.title',
    action: homePath()
  }

  const component = shallow(
    <ThemeProvider theme={{} as any}>
      <Item {...props}/>
    </ThemeProvider>
  )
  expect(snapshot(component)).toMatchSnapshot()
})