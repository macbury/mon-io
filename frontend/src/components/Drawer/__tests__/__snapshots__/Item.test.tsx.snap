// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<ContextProvider
  value={Object {}}
>
  <Item
    action={
      Object {
        "routeName": "home",
        "type": "Navigation/NAVIGATE",
      }
    }
    icon="settings"
    titleKey="pages.series.title"
  />
</ContextProvider>
`;
